### ISSUE TITLE
Resource test harness

### ISSUE THREAD

**Issue by bethge** [rebuy-de/aws-nuke#152]
The more edge cases we run into, the stronger the argument for integration tests. Even if it is rather involved to bootstrap the resource and validate deletion, the effort would even in a rather short-term outweigh the cost, especially for popular AWS services like S3.

@svenwltr and I have been whiteboarding a bit and would:
- use terraform to create the resource under test. We would add a `<resource>.tf` file along every `<resource>.go` with the resource declaration and a `<resource>_test.go` file for test execution.
- nuke it with `--target` flag
- Validate deletion of the resource via output of `terraform plan`

I would stitch together a prototype to see if this is a viable solution.

Very open to suggestions.

**Comment by tomvachon** [comment]
It could become very costly depending on some of the potential types I'm adding to get 100% coverage.

While I think the TF approach has some validity, how about also adding in GoSpec for wider coverage.  The docs on how to contribute and guidelines are needed before we can do any of this though.

**Comment by svenwltr** [comment]
> It could become very costly depending on some of the potential types I'm adding to get 100% coverage.

The question is if we actually should aim to 100% coverage, because I think this will be almost impossible to achieve.

> While I think the TF approach has some validity, how about also adding in GoSpec for wider coverage. 

What do you mean with "GoSpec"?

**Comment by svenwltr** [comment]
I don't know if there is any `rpec` like tooling in Golang, but the included unit test framework is already very good (IMHO). However, I don't think unit tests make much sense for the actual resources (ie everything in 
the `resources/` directory), because the resource are well isolated and quite easy to write. The most common things which can happen here are syntax/type errors, which are covered by the compiling and logical errors which are API related. The later ones a very hard to detect without using the actual API.

Just take a look at the problems which popped up in the recent weeks. I don't think any of these would be covered by unit tests.

**Comment by tomvachon** [comment]
Ok, sounds good then

**Comment by svenwltr** [comment]
A first simple test could be to let *aws-nuke* run on an empty account on all regions and expect no errors.

**Comment by eschwartz** [comment]
@bethge @tomvachon @svenwltr -- is there still any appetite for integration tests against AWS resources here? I have some time allotted to working on this repo, and I'd be very happy to put together a proof-of-concept for what an integration test might be look like.

There are a few ways we could approach this, but I would probably try to keep the tests themselves as simple as possible for now, to prove it out -- eg, use the AWS SDK to create some resources, then use the aws-nuke resource packages to scan/filter/remove the resources.

The trickier question is where and when do these tests run? Can we create a new AWS account with creds in TravisCI? Do we run them on every PR, or just on master? How do we control the AWS account, so open-source contributors can't spin up bitcoin miners or whatever in the test account?

**Comment by svenwltr** [comment]
Hello @eschwartz,

we are still interested in this. There was also an [PR a long time ago](https://github.com/rebuy-de/aws-nuke/pull/177), but we did not have time to finalize it.

Using the AWS SDK sounds fine for an initial test, but it might get cumbersome for some resources (ie resources that need a lot of other resource beforehand, like EKS). That's why we started with Terraform. On the other hand it might be beneficial to start with a simple approach first.

We can create an AWS account for regularly testing the master, but we cannot test every PR with this, because it would basically allow arbitrary code execution in our account. The account and nightly build setup would be on our hand. Therefore the tests should be designed that everyone can run the test for a specific resource in their own account.
